vowpalwabbit/global_data.cc

```diff
@@ -167,7 +167,20 @@ void vw::learn(example& ec)
   if (ec.test_only || !training)
     VW::LEARNER::as_singleline(l)->predict(ec);
   else
-    VW::LEARNER::as_singleline(l)->learn(ec);
+  {
+    if (l->predict_before_learn)
```

Comment:
The goal for this PR was to introduce predict/learn semantics and switch to it where possible.  Initially this covered many more reductions but due to efficiency reasons this had to be rolled back.  This flag was introduced to address the efficiency issues.

Next steps:
1) All new reductions from now on should use the new semantics
2) There are some high value changes around library mode (predict file, progressive validation) that can take advantage of new semantics.  Along with these changes, I expect more reductions to migrate to new semantics.  The exact order will depend on prioritization and scheduling.
3) Other reductions beyond what's in step (2) will have to be addressed depending on priority.

Reply:
Can you answer the first question though? was that intended?

Reply:
Yes.  This is intended.  Predictions in top-most reduction are used in progressive validation, output to screen, written to predict files so always needed.

Reply:
wouldn't this imply that if you have a stack with the following:
`reduction(predict_before_learn=True) -> reduction(predict_before_learn=False) -> ... -> base_reduction()` you would effectively add an extra call to predict with zero benefit?

Reply:
A reduction with predict_before_learn == True supports new semantics for predict/learn.  If the framework calls predict before learn, the top level reduction will likely induce calls to predict all the way down the chain.  Since the framework calls learn after predict, it's possible that one of reductions down the chain makes extra calls to predict.  However, caching predictions downstream fails because many calls to base predict from reductions deeper in the stack.  As a rough experiment, I attempted to cache predictions in gd.  This failed because reductions can modify examples and make predict calls to the same learner.
The solution to more efficient learn is to a) detect conditions where predict is not needed for learn b) let the framework avoid base level predict call.